File: automator/src/main/java/com/stardust/automator/UiObjectCollection.java
package com.stardust.automator;

import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;

import com.stardust.util.Consumer;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;
import java.util.Spliterator;
import java.util.function.Predicate;
import java.util.function.UnaryOperator;
import java.util.stream.Stream;

import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ACCESSIBILITY_FOCUS;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_COLUMN_INT;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_PROGRESS_VALUE;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_ROW_INT;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_SELECTION_END_INT;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_SELECTION_START_INT;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_CLEAR_ACCESSIBILITY_FOCUS;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_CLEAR_FOCUS;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_CLICK;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_COLLAPSE;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_COPY;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_CUT;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_DISMISS;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_EXPAND;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_FOCUS;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_LONG_CLICK;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_PASTE;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_SCROLL_BACKWARD;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_SCROLL_FORWARD;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_SELECT;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_SET_SELECTION;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.ACTION_SET_TEXT;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_CONTEXT_CLICK;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_DOWN;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_LEFT;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_RIGHT;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_TO_POSITION;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_UP;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SET_PROGRESS;
import static android.support.v4.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SHOW_ON_SCREEN;

/**
 * Created by Stardust on 2017/3/9.
 */

public class UiObjectCollection {

    public static final UiObjectCollection EMPTY = UiObjectCollection.of(Collections.<UiObject>emptyList());

    public static UiObjectCollection of(List<UiObject> list) {
        return new UiObjectCollection(list);
    }

    private List<UiObject> mNodes;

    private UiObjectCollection(List<UiObject> list) {
        mNodes = list;
    }

    public UiObject[] toArray() {
        return mNodes.toArray(new UiObject[0]);
    }

    public <T> T[] toArray(T[] a) {
        return mNodes.toArray(a);
    }

    public boolean add(UiObject uiObject) {
        return mNodes.add(uiObject);
    }

    public boolean remove(Object o) {
        return mNodes.remove(o);
    }

    public boolean containsAll(Collection<?> c) {
        return mNodes.containsAll(c);
    }


    public boolean addAll(Collection<? extends UiObject> c) {
        return mNodes.addAll(c);
    }


    public boolean addAll(int index, Collection<? extends UiObject> c) {
        return mNodes.addAll(index, c);
    }


    public boolean removeAll(Collection<?> c) {
        return mNodes.removeAll(c);
    }


    public boolean retainAll(Collection<?> c) {
        return mNodes.retainAll(c);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)

    public void replaceAll(UnaryOperator<UiObject> operator) {
        mNodes.replaceAll(operator);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)

    public void sort(Comparator<? super UiObject> c) {
        mNodes.sort(c);
    }


    public void clear() {
        mNodes.clear();
    }

    public boolean performAction(int action) {
        boolean fail = false;
        for (UiObject node : mNodes) {
            if (!node.performAction(action)) {
                fail = true;
            }
        }
        return !fail;
    }

    public boolean performAction(int action, ActionArgument... arguments) {
        boolean fail = false;
        Bundle bundle = argumentsToBundle(arguments);
        for (UiObject node : mNodes) {
            if (!node.performAction(action, bundle)) {
                fail = true;
            }
        }
        return !fail;
    }

    private Bundle argumentsToBundle(ActionArgument[] arguments) {
        Bundle bundle = new Bundle();
        for (ActionArgument arg : arguments) {
            arg.putIn(bundle);
        }
        return bundle;
    }

    public boolean click() {
        return performAction(ACTION_CLICK);
    }

    public boolean longClick() {
        return performAction(ACTION_LONG_CLICK);
    }

    public boolean accessibilityFocus() {
        return performAction(ACTION_ACCESSIBILITY_FOCUS);
    }

    public boolean clearAccessibilityFocus() {
        return performAction(ACTION_CLEAR_ACCESSIBILITY_FOCUS);
    }

    public boolean focus() {
        return performAction(ACTION_FOCUS);
    }

    public boolean clearFocus() {
        return performAction(ACTION_CLEAR_FOCUS);
    }

    public boolean copy() {
        return performAction(ACTION_COPY);
    }

    public boolean paste() {
        return performAction(ACTION_PASTE);
    }

    public boolean select() {
        return performAction(ACTION_SELECT);
    }

    public boolean cut() {
        return performAction(ACTION_CUT);
    }

    public boolean collapse() {
        return performAction(ACTION_COLLAPSE);
    }

    public boolean expand() {
        return performAction(ACTION_EXPAND);
    }

    public boolean dismiss() {
        return performAction(ACTION_DISMISS);
    }

    public boolean show() {
        return performAction(ACTION_SHOW_ON_SCREEN.getId());
    }

    public boolean scrollForward() {
        return performAction(ACTION_SCROLL_FORWARD);
    }

    public boolean scrollBackward() {
        return performAction(ACTION_SCROLL_BACKWARD);
    }

    public boolean scrollUp() {
        return performAction(ACTION_SCROLL_UP.getId());
    }

    public boolean scrollDown() {
        return performAction(ACTION_SCROLL_DOWN.getId());
    }

    public boolean scrollLeft() {
        return performAction(ACTION_SCROLL_LEFT.getId());
    }

    public boolean scrollRight() {
        return performAction(ACTION_SCROLL_RIGHT.getId());
    }

    public boolean contextClick() {
        return performAction(ACTION_CONTEXT_CLICK.getId());
    }

    public boolean setSelection(int s, int e) {
        return performAction(ACTION_SET_SELECTION,
                new ActionArgument.IntActionArgument(ACTION_ARGUMENT_SELECTION_START_INT, s),
                new ActionArgument.IntActionArgument(ACTION_ARGUMENT_SELECTION_END_INT, e));
    }

    public boolean setText(CharSequence text) {
        return performAction(ACTION_SET_TEXT,
                new ActionArgument.CharSequenceActionArgument(ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE, text));
    }

    public boolean setProgress(float value) {
        return performAction(ACTION_SET_PROGRESS.getId(),
                new ActionArgument.FloatActionArgument(ACTION_ARGUMENT_PROGRESS_VALUE, value));

    }

    public boolean scrollTo(int row, int column) {
        return performAction(ACTION_SCROLL_TO_POSITION.getId(),
                new ActionArgument.IntActionArgument(ACTION_ARGUMENT_ROW_INT, row),
                new ActionArgument.IntActionArgument(ACTION_ARGUMENT_COLUMN_INT, column));
    }

    public UiObject get(int i) {
        return mNodes.get(i);
    }


    public UiObject set(int index, UiObject element) {
        return mNodes.set(index, element);
    }


    public void add(int index, UiObject element) {
        mNodes.add(index, element);
    }


    public UiObject remove(int index) {
        return mNodes.remove(index);
    }


    public int indexOf(Object o) {
        return mNodes.indexOf(o);
    }


    public int lastIndexOf(Object o) {
        return mNodes.lastIndexOf(o);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)

    public ListIterator<UiObject> listIterator() {
        return mNodes.listIterator();
    }

    @RequiresApi(api = Build.VERSION_CODES.N)

    public ListIterator<UiObject> listIterator(int index) {
        return mNodes.listIterator(index);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)

    public List<UiObject> subList(int fromIndex, int toIndex) {
        return mNodes.subList(fromIndex, toIndex);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)

    public Spliterator<UiObject> spliterator() {
        return mNodes.spliterator();
    }

    @RequiresApi(api = Build.VERSION_CODES.N)

    public boolean removeIf(Predicate<? super UiObject> filter) {
        return mNodes.removeIf(filter);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)

    public Stream<UiObject> stream() {
        return mNodes.stream();
    }

    @RequiresApi(api = Build.VERSION_CODES.N)

    public Stream<UiObject> parallelStream() {
        return mNodes.parallelStream();
    }

    public int size() {
        return mNodes.size();
    }


    public boolean isEmpty() {
        return mNodes.isEmpty();
    }


    public boolean contains(Object o) {
        return mNodes.contains(o);
    }


    public Iterator<UiObject> iterator() {
        return mNodes.iterator();
    }

    public UiObjectCollection each(Consumer<UiObject> consumer) {
        for (UiObject uiObject : mNodes) {
            consumer.accept(uiObject);
        }
        return this;
    }

    public UiObjectCollection find(UiGlobalSelector selector) {
        List<UiObject> list = new ArrayList<>();
        for (UiObject object : mNodes) {
            list.addAll(selector.findOf(object).mNodes);
        }
        return of(list);
    }

    @Nullable
    public UiObject findOne(UiGlobalSelector selector) {
        for (UiObject object : mNodes) {
            UiObject result = selector.findOneOf(object);
            if (result != null)
                return result;
        }
        return null;
    }

    public boolean empty() {
        return size() == 0;
    }

    public boolean nonEmpty() {
        return size() != 0;
    }

}

File: automator/src/main/java/com/stardust/automator/UiObjectCollection.kt
package com.stardust.automator

import android.os.Bundle

import com.stardust.util.Consumer

import java.util.ArrayList
import java.util.Collections

import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ACCESSIBILITY_FOCUS
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_COLUMN_INT
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_PROGRESS_VALUE
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_ROW_INT
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_SELECTION_END_INT
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_SELECTION_START_INT
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_CLEAR_ACCESSIBILITY_FOCUS
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_CLEAR_FOCUS
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_CLICK
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_COLLAPSE
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_COPY
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_CUT
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_DISMISS
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_EXPAND
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_FOCUS
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_LONG_CLICK
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_PASTE
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_SCROLL_BACKWARD
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_SCROLL_FORWARD
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_SELECT
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_SET_SELECTION
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.ACTION_SET_TEXT
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_CONTEXT_CLICK
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_DOWN
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_LEFT
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_RIGHT
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_TO_POSITION
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_UP
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SET_PROGRESS
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SHOW_ON_SCREEN

/**
 * Created by Stardust on 2017/3/9.
 */

class UiObjectCollection private constructor(private val mNodes: List<UiObject?>) {


    val isEmpty: Boolean
        get() = mNodes.isEmpty()

    fun toArray(): Array<UiObject?> {
        return mNodes.toTypedArray()
    }

    fun performAction(action: Int): Boolean {
        var fail = false
        for (node in mNodes) {
            val succeed = node?.performAction(action) ?: false
            if (!succeed) {
                fail = true
            }
        }
        return !fail
    }

    fun performAction(action: Int, vararg arguments: ActionArgument): Boolean {
        var fail = false
        val bundle = argumentsToBundle(arguments)
        for (node in mNodes) {
            val succeed = node?.performAction(action, bundle) ?: false
            if (succeed) {
                fail = true
            }
        }
        return !fail
    }

    private fun argumentsToBundle(arguments: Array<out ActionArgument>): Bundle {
        val bundle = Bundle()
        for (arg in arguments) {
            arg.putIn(bundle)
        }
        return bundle
    }

    fun click(): Boolean {
        return performAction(ACTION_CLICK)
    }

    fun longClick(): Boolean {
        return performAction(ACTION_LONG_CLICK)
    }

    fun accessibilityFocus(): Boolean {
        return performAction(ACTION_ACCESSIBILITY_FOCUS)
    }

    fun clearAccessibilityFocus(): Boolean {
        return performAction(ACTION_CLEAR_ACCESSIBILITY_FOCUS)
    }

    fun focus(): Boolean {
        return performAction(ACTION_FOCUS)
    }

    fun clearFocus(): Boolean {
        return performAction(ACTION_CLEAR_FOCUS)
    }

    fun copy(): Boolean {
        return performAction(ACTION_COPY)
    }

    fun paste(): Boolean {
        return performAction(ACTION_PASTE)
    }

    fun select(): Boolean {
        return performAction(ACTION_SELECT)
    }

    fun cut(): Boolean {
        return performAction(ACTION_CUT)
    }

    fun collapse(): Boolean {
        return performAction(ACTION_COLLAPSE)
    }

    fun expand(): Boolean {
        return performAction(ACTION_EXPAND)
    }

    fun dismiss(): Boolean {
        return performAction(ACTION_DISMISS)
    }

    fun show(): Boolean {
        return performAction(ACTION_SHOW_ON_SCREEN.id)
    }

    fun scrollForward(): Boolean {
        return performAction(ACTION_SCROLL_FORWARD)
    }

    fun scrollBackward(): Boolean {
        return performAction(ACTION_SCROLL_BACKWARD)
    }

    fun scrollUp(): Boolean {
        return performAction(ACTION_SCROLL_UP.id)
    }

    fun scrollDown(): Boolean {
        return performAction(ACTION_SCROLL_DOWN.id)
    }

    fun scrollLeft(): Boolean {
        return performAction(ACTION_SCROLL_LEFT.id)
    }

    fun scrollRight(): Boolean {
        return performAction(ACTION_SCROLL_RIGHT.id)
    }

    fun contextClick(): Boolean {
        return performAction(ACTION_CONTEXT_CLICK.id)
    }

    fun setSelection(s: Int, e: Int): Boolean {
        return performAction(ACTION_SET_SELECTION,
                ActionArgument.IntActionArgument(ACTION_ARGUMENT_SELECTION_START_INT, s),
                ActionArgument.IntActionArgument(ACTION_ARGUMENT_SELECTION_END_INT, e))
    }

    fun setText(text: CharSequence): Boolean {
        return performAction(ACTION_SET_TEXT,
                ActionArgument.CharSequenceActionArgument(ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE, text))
    }

    fun setProgress(value: Float): Boolean {
        return performAction(ACTION_SET_PROGRESS.id,
                ActionArgument.FloatActionArgument(ACTION_ARGUMENT_PROGRESS_VALUE, value))

    }

    fun scrollTo(row: Int, column: Int): Boolean {
        return performAction(ACTION_SCROLL_TO_POSITION.id,
                ActionArgument.IntActionArgument(ACTION_ARGUMENT_ROW_INT, row),
                ActionArgument.IntActionArgument(ACTION_ARGUMENT_COLUMN_INT, column))
    }

    operator fun get(i: Int): UiObject? {
        return mNodes[i]
    }

    fun indexOf(o: UiObject?): Int {
        return mNodes.indexOf(o)
    }


    fun lastIndexOf(o: UiObject?): Int {
        return mNodes.lastIndexOf(o)
    }

    fun size(): Int {
        return mNodes.size
    }


    operator fun contains(o: UiObject?): Boolean {
        return mNodes.contains(o)
    }

    operator fun iterator(): Iterator<UiObject?> {
        return mNodes.iterator()
    }

    fun each(consumer: Consumer<UiObject>): UiObjectCollection {
        for (uiObject in mNodes) {
            consumer.accept(uiObject)
        }
        return this
    }

    fun find(selector: UiGlobalSelector): UiObjectCollection {
        val list = ArrayList<UiObject?>()
        for (`object` in mNodes) {
            `object`?.let {
                list.addAll(selector.findOf(it).mNodes)
            }
        }
        return of(list)
    }

    fun findOne(selector: UiGlobalSelector): UiObject? {
        for (`object` in mNodes) {
            val result = `object`?.run { selector.findOneOf(this) }
            if (result != null)
                return result
        }
        return null
    }

    fun empty(): Boolean {
        return size() == 0
    }

    fun nonEmpty(): Boolean {
        return size() != 0
    }

    companion object {

        val EMPTY = UiObjectCollection.of(emptyList())

        fun of(list: List<UiObject?>): UiObjectCollection {
            return UiObjectCollection(list)
        }
    }

}
